tools/base/tests/test_functional.py

```diff
@@ -0,0 +1,81 @@
+from unittest.mock import AsyncMock
+
+import pytest
+
+from tools.base import functional
+
+
+@pytest.mark.asyncio
+@pytest.mark.parametrize("cache", [None, True, False])
+@pytest.mark.parametrize("raises", [True, False])
+async def test_functional_async_property(cache, raises):
```

Comment:
this is probs more of an integration test than a unit test but as its reaching into the bowels of python im kinda happy with that